Mathlib/Topology/Category/Profinite/Nobeling.lean

```diff
@@ -0,0 +1,1549 @@
+    exact Submodule.subset_span ⟨a, ⟨h',rfl⟩⟩
+  · rfl
+  · exfalso; apply ha; rw [h]
+    exact Submodule.subset_span ⟨b, ⟨h',rfl⟩⟩
+
+/-- The image of the good products in the module `LocallyConstant C ℤ`. -/
+def range := Set.range (GoodProducts.eval C)
+
+/-- The type of good products is equivalent to its image. -/
+noncomputable
+def equiv_range : GoodProducts C ≃ range C :=
+  Equiv.ofInjective (eval C) (injective C)
+
+theorem equiv_toFun_eq_eval : (equiv_range C).toFun = Set.rangeFactorization (eval C) := by rfl
+
+theorem linearIndependent_iff_range : LinearIndependent ℤ (GoodProducts.eval C) ↔
+    LinearIndependent ℤ (fun (p : range C) ↦ p.1) := by
+  rw [← @Set.rangeFactorization_eq _ _ (GoodProducts.eval C), ← equiv_toFun_eq_eval C]
+  exact linearIndependent_equiv (equiv_range C)
+
+end GoodProducts
+
+namespace Products
+
+theorem eval_eq (l : Products I) (x : C) :
+    l.eval C x = if ∀ i, i ∈ l.val → (x.val i = true) then 1 else 0 := by
+  change LocallyConstant.evalMonoidHom x (l.eval C) = _
+  rw [eval, map_list_prod]
+  split_ifs with h
+  · simp only [List.map_map]
+    apply List.prod_eq_one
+    simp only [List.mem_map, Function.comp_apply]
+    rintro _ ⟨i, hi, rfl⟩
+    exact if_pos (h i hi)
+  · simp only [List.map_map, List.prod_eq_zero_iff, List.mem_map, Function.comp_apply]
+    push_neg at h
+    convert h with i
+    dsimp [LocallyConstant.evalMonoidHom, e]
+    simp only [ite_eq_right_iff]
+
+theorem evalFacProp {l : Products I} (J : I → Prop)
+    (h : ∀ a, a ∈ l.val → J a) [∀ j, Decidable (J j)] :
+    l.eval (π C J) ∘ ProjRestrict C J = l.eval C := by
+  ext x
+  dsimp [ProjRestrict]
+  rw [Products.eval_eq, Products.eval_eq]
+  congr
+  apply forall_congr; intro i
+  apply forall_congr; intro hi
+  simp [h i hi, Proj]
+
+theorem evalFacProps {l : Products I} (J K : I → Prop)
+    (h : ∀ a, a ∈ l.val → J a) [∀ j, Decidable (J j)] [∀ j, Decidable (K j)]
+    (hJK : ∀ i, J i → K i) :
+    l.eval (π C J) ∘ ProjRestricts C hJK = l.eval (π C K) := by
+  have : l.eval (π C J) ∘ Homeomorph.setCongr (proj_eq_of_subset C J K hJK) =
+      l.eval (π (π C K) J) := by
+    ext; simp [Homeomorph.setCongr, Products.eval_eq]
+  rw [ProjRestricts, ← Function.comp.assoc, this, ← evalFacProp (π C K) J h]
+
+theorem prop_of_isGood  {l : Products I} (J : I → Prop) [∀ j, Decidable (J j)]
+    (h : l.isGood (π C J)) : ∀ a, a ∈ l.val → J a := by
+  intro i hi
+  by_contra h'
+  apply h
+  suffices : eval (π C J) l = 0
+  · rw [this]
+    exact Submodule.zero_mem _
+  ext ⟨_, ⟨_, ⟨_, rfl⟩⟩⟩
+  rw [eval_eq, if_neg]
+  · rfl
+  · intro h
+    specialize h i hi
+    simp only [Proj, Bool.ite_eq_true_distrib, if_false_right_eq_and] at h
+    exact h' h.1
+
+instance : IsWellFounded (Products I) (·<·) := by
+  have : (· < · : Products I → _ → _) = (fun l m ↦ List.Lex (·<·) l.val m.val) := by
+    ext; exact lt_iff_lex_lt _ _
+  rw [this]
+  dsimp [Products]
+  rw [(by rfl : (·>· : I → _) = flip (·<·))]
+  infer_instance
```

Comment:
I think this instance should go straight after `lt_iff_lex_lt` on line 242, rather than here.